### PR TITLE
docs: contributing

### LINKED ISSUE #58: Create Contributing.md
## Where?
In the root directory of the project.

## Who?
New developers that want to contribute to this project

## What?
 - How to use SyMPC with Syft
 - Installation steps and testing steps for new changes

## Additional Context
Nothing more

### PR DESCRIPTION
**Pull request by danielorihuela**
## Description
Copied and modified the contributing documentation from PySyft.
Explains how to install and test SyMPC.

Closes #58 

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: CONTRIBUTING.md
Comment (aanurraj) [inline CONTRIBUTING.md:35]: this is not linked.

File: CONTRIBUTING.md
Comment (aanurraj) [inline CONTRIBUTING.md:43]: this too is not linked.

File: CONTRIBUTING.md
Comment (aanurraj) [inline CONTRIBUTING.md:170]: though it's just an example can you replace this with python 3.9 ? also we don't even test it for 3.6

File: CONTRIBUTING.md
Comment (aanurraj) [inline CONTRIBUTING.md:253]: extra \ here.

**Review by aanurraj — CHANGES_REQUESTED** [review]
Looks good, I requested few changes you might need to do that. Also, have you checked all the steps if they are working correctly?

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md]: thanks!

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md]: thanks!

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md]: Done, I also changed some places where it was told that python 3.6 is supported to python 3.7, which is the minimum version that is automatically tested.

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md:251]: I am not sure it is extra. If I remove it, the Markdown visualizer of visual code will show `https://github.com//SyMPC` instead of `https://github.com/<your-username>/SyMPC`

**Comment by danielorihuela** [comment]
> Looks good, I requested few changes you might need to do that. Also, have you checked all the steps if they are working correctly?

I checked all expect the windows and macos parts, I only have a linux computer.
Also, there is a problem with the requirements.txt and requirements.dev.txt when you try to install them. But I guess that is due to the transformation of pysyft main repo to a monorepo.

**Comment by gmuraru** [comment]
> > Looks good, I requested few changes you might need to do that. Also, have you checked all the steps if they are working correctly?
> 
> I checked all except the windows and macos parts, I only have a linux computer.
> Also, there is a problem with the requirements.txt and requirements.dev.txt when you try to install them. But I guess that is due to the transformation of pysyft main repo to a monorepo.

It should work now.

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:3]: Q: Should we remove this?

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:316]: This should be changed (remove the username and add `...`)

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:385]: Q: I think SyMPC depends on PySyft, right?

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:397]: `cd PySyft/packages/syft` ?

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:399]: I think you need to only do:
```pip install -e PySyft/packages/syft``` (no need to use the requirements)

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:413]: 3 and 4 is optional - because if you already have the SyMPC and PySyft repos locally and you did the previous step, you don't need to do this every time you make a change.

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:437]: NITPICK: I think you should run only `pytest -n auto tests` (We do not have slow vs fast tests -- this is for Syft to have faster local development)

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:458]: @jmaunon we do this in the `pre-commit hooks`, right?

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:489]: Q: Is this working? (@jmaunon )

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:74]: NITPICK: Leave only Linux :D (could you update the examples to remove the "brew")
We use `torchcsprng` that (I think) it is only available on Linux)

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:84]: This should be removed if the above is true

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:96]: This should be removed if the above is true

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:111]: This should be removed if the above is true

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:121]: This should be removed if the above is true

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:129]: For Linux only

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:178]: python.exe should be removed

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:540]: :+1:

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:498]: Remove the install - we do this in `requirements.dev.txt` (we need only the `pre-commit install` I think)

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:325]: Remove this

**Review by gmuraru — CHANGES_REQUESTED** [review]
I will give tomorrow another look - but added some comments

File: CONTRIBUTING.md
Comment (gmuraru) [reply CONTRIBUTING.md:540]: `code_sympc` is only for repo-related updates - could you remove it?

File: CONTRIBUTING.md
Comment (rasswanth-s) [reply CONTRIBUTING.md:74]: I looked at the docs,``` torchcsprng ``` supports Windows(version>=0.2.0) .It supports mac only on CPU but not on CUDA. https://pypi.org/project/torchcsprng/

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md:458]: If I am not wrong, I get the information from this file.

image

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md:3]: I am not sure, as you know I copied the structure and lot of parts from PySyft. I do not know if its nice to let it here for context. As you prefer, there is a link in the `Linking the PySyft src` section. So I think, it should be safe.

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md:74]: Maybe, I should remove those parts and tell that torchspring can be use in windows what currentyle we do not provide any kind of support?

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md:488]: This is actually a command that I "invented" myself. On my computer worked at least. You should be able to execute it too.

File: CONTRIBUTING.md
Comment (danielorihuela) [reply CONTRIBUTING.md]: I say it explicitly now. Is it better now?

File: CONTRIBUTING.md
Comment (kamathhrishi) [reply CONTRIBUTING.md:75]: Hey @rasswanth-s, Mac does not support CUDA so even Pytorch and torchcspring does'nt support cuda on Mac.

Comment (rasswanth-s) [reply CONTRIBUTING.md:75]: Then I think, it would work on all three platforms.

File: CONTRIBUTING.md
Comment (jmaunon) [reply CONTRIBUTING.md:490]: the only thing is needed to work is to install sphinx

**Review by gmuraru — APPROVED** [review]
LGTM!

File: CONTRIBUTING.md
Comment (gmuraru) [inline CONTRIBUTING.md:11]: Q: Could you check if `sycret` ahs support for Windows and MacOS?

Comment (danielorihuela) [reply CONTRIBUTING.md:11]: I checked https://github.com/OpenMined/sycret and it does not mention it. Rust can be installed in all systems so I suppose there should not be any problem (source: https://forge.rust-lang.org/infra/other-installation-methods.html). Maybe @tholop can give us a definitive answer.

Also, before going to master @gmuraru. Should I remove all the macos and windows parts that you told me to in the reviews? Should I remove the initial part where we say that we follow PySyft contributions?

Comment (gmuraru) [reply CONTRIBUTING.md:11]: No no, if `torchcsprang` is also available there, I think we are good for the moment.

Comment (tholop) [reply CONTRIBUTING.md:11]: @danielorihuela @gmuraru I tested Sycret on Linux and MacOS. It should work on Windows but I never tried. 
How are you guys testing SyMPC on Windows? Do you have a special workflow for that (I couldn't find it) or do you spam someone who is running Windows?

Comment (gmuraru) [reply CONTRIBUTING.md:11]: We do not have actions for Windows and OSX. @danielorihuela could you add ```Windows``` and ```MacOS``` in the `.github` workflow, for PR-tests and lets see if they work

Comment (danielorihuela) [reply CONTRIBUTING.md:11]: Okay, I am gonna give it a try.

Comment (tholop) [reply CONTRIBUTING.md:11]: Thanks @danielorihuela! If it works I'll do the same for Sycret :slightly_smiling_face:

Comment (gmuraru) [reply CONTRIBUTING.md:11]: I think you only need to change this to have multiple values? `runs-on: ubuntu-latest` (the PySyft team is doing this)

Comment (danielorihuela) [reply CONTRIBUTING.md:11]: Yeah I saw, thanks!

I made it work yesterday with MacOS. The problem is with windows. It does not have any command "hash -r", so I removed it. But then the trilom github action fails. I do not really know if it is related or not.

image

Comment (gmuraru) [reply CONTRIBUTING.md:11]: Could you also rebase, I made a change to one of the workflow files to have a new version for `trilom`

Comment (danielorihuela) [reply CONTRIBUTING.md:11]: Yeah, I tried that yesterday too. 
image

Comment (danielorihuela) [reply CONTRIBUTING.md:11]: Maybe this issue has something to do https://github.com/trilom/file-changes-action/issues/112

Comment (danielorihuela) [reply CONTRIBUTING.md:11]: @gmuraru I could make the tests works by changing the github action to get all the changed files (jitterbit/get-changed-files@v1). You can see a run https://github.com/danielorihuela/SyMPC/pull/12/checks?check_run_id=2632889341.

Comment (gmuraru) [reply CONTRIBUTING.md:11]: Yep! It looks good with the change from there

**Comment by gmuraru** [comment]
Only solve the conflict and I think this is good to go :rocket:

**Comment by danielorihuela** [comment]
All tests passed but is waiting for python-tests (3.8). Is that a bug?

**Comment by gmuraru** [comment]
> All tests passed but is waiting for python-tests (3.8). Is that a bug?

I updated the branch. I am not sure...I have seen it hold up one other time - let's see them now

**Comment by gmuraru** [comment]
I will force squash and merge this after the tests are passing

**Comment by danielorihuela** [comment]
> I will force squash and merge this after the tests are passing

As you wish, but I would not like that this behaviour gets expanded to all PRs.

**Comment by gmuraru** [comment]
Asked @aanurraj to look again over this - I think it is good to go, but it is better to have another pair of eyes to look over this :eyes:

**Comment by gmuraru** [comment]
> > I will force squash and merge this after the tests are passing
> 
> As you wish, but I would not like that this behaviour gets expanded to all PRs. Are you sure this will disappear?
> 
> It is curious, I have done a run in my fork and that bug does not happen.
> [danielorihuela#13 (comment)](https://github.com/danielorihuela/SyMPC/pull/13#partial-pull-merging)

Yeah...we should keep an eye on this problem...and if this happens once more, we should investigate this.

**Review by aanurraj — APPROVED** [review]

**Pull request merged**
